### PR TITLE
Delay beefy worker initialization while network is on major sync

### PR DESCRIPTION
**Pull request by Wizdave97**
Fixes https://github.com/paritytech/grandpa-bridge-gadget/issues/112

### CODE REVIEW

File: client/beefy/src/worker.rs [hunk 1]
```diff
@@ -400,6 +408,11 @@ where
 		));
 
 		loop {
+			if self.sync_oracle.is_major_syncing() {
+				debug!(target: "beefy", "Skipping initialization due to sync.");
+				continue
+			}
+
```
Comment (seunlanlege) [inline client/beefy/src/worker.rs:415]: I feel like this log would be a bit too noisy, is there anyway we can make a function `wait_for_major_syncing` that returns a future that returns when major syncing ends?

```suggestion
			if self.sync_oracle.is_major_syncing() {
				debug!(target: "beefy", "Skipping initialization due to sync.");
				self.sync_oracle..wait_for_major_syncing().await;
			}

```

File: client/beefy/src/worker.rs [hunk 2]
```diff
@@ -400,6 +409,11 @@ where
 		));
 
 		loop {
+			if self.sync_oracle.is_major_syncing() {
+				debug!(target: "beefy", "Skipping initialization due to sync.");
```
Comment (acatangiu) [inline client/beefy/src/worker.rs:413]: This is not necessarily initialization, it is the main loop of the worker. I would rephrase to smth like:

```suggestion
				debug!(target: "beefy", "Waiting for major sync to complete.");
```

File: client/beefy/src/lib.rs [hunk 3]
```diff
@@ -106,13 +107,14 @@ where
...
 /// BEEFY gadget initialization parameters.
-pub struct BeefyParams<B, BE, C, N>
+pub struct BeefyParams<B, BE, C, N, SO>
 where
 	B: Block,
 	BE: Backend<B>,
 	C: Client<B, BE>,
 	C::Api: BeefyApi<B>,
 	N: GossipNetwork<B> + Clone + Send + 'static,
+	SO: SyncOracle + Send + Sync + Clone + 'static,
```
Comment (acatangiu) [inline client/beefy/src/lib.rs:117]: Would these two ever be provided by different objects? If not, I think it would make sense to simply squash them together, they're both mandatory anyway...

**Review by acatangiu — APPROVED** [review]

File: client/beefy/src/worker.rs [hunk 4]
```diff
@@ -430,6 +444,19 @@ where
...
+fn wait_for_major_syncing<SO: SyncOracle + Send + Sync + Clone + 'static>(
+	mut sync_oracle: SO,
+) -> impl future::Future<Output = ()> {
+	return future::poll_fn(move |_cx| {
+		if sync_oracle.is_major_syncing() {
+			Poll::Pending
+		} else {
+			Poll::Ready(())
+		}
+	})
```
Comment (seunlanlege) [inline client/beefy/src/worker.rs:458]: this function has no way of knowing when the sync is over, i suggest adding it to the  sync_oracle itself

Comment (Wizdave97) [reply client/beefy/src/worker.rs:458]: What if it's rescheduled for polling with `cx.waker().wake_by_ref()` before returning pending, Would that be fine?

File: client/beefy/src/worker.rs [hunk 5]
```diff
@@ -430,6 +444,20 @@ where
...
 }
 
+/// Returns a future that waits until major sync is done
+/// before completing
+fn wait_for_major_syncing<SO: SyncOracle + Send + Sync + Clone + 'static>(
+	mut sync_oracle: SO,
+) -> impl future::Future<Output = ()> {
+	return future::poll_fn(move |cx| {
+		if sync_oracle.is_major_syncing() {
+			cx.waker().wake_by_ref();
```
Comment (acatangiu) [inline client/beefy/src/worker.rs:454]: this is equivalent to a busy-wait loop, right? Like @seunlanlege said, this future needs to be tied to the sync_oracle waker.

**Review by acatangiu — CHANGES_REQUESTED** [review]

**Comment by seunlanlege** [comment]
Bump

**Comment by acatangiu** [comment]
Alternative PR https://github.com/paritytech/substrate/pull/10882 merged
